### ISSUE TITLE
Allow comment of a "else[ if]" block to be dedented

### ISSUE THREAD

**Issue by clementprevot** [beautify-web/js-beautify#1069]
# Description
When writing `if...else` block with 1TBS style, it's hard to comment the `else` (or `else if`) comment as the comment for the `else` block is indented at the same level than the `if` block content.

# Input
The code looked like this before beautification:
```
// In this case
if (foo) {
    // Do something
// Else do something else
} else {
    // Do something else
}
```

# Expected Output
The code should have looked like this after beautification:
```
if (foo) {
    // Do something
// Else do something else
} else {
    // Do something else
}
```

# Actual Output
The  code actually looked like this after beautification:
```
if (foo) {
    // Do something
    // Else do something else
} else {
    // Do something else
}
```

## Environment
OS: Linux Mint 17

## Settings
Example:
```json
{
    "indent_size": 4,
    "indent_char": " ",
    "eol": "\n",
    "indent_level": 0,
    "indent_with_tabs": false,
    "preserve_newlines": true,
    "max_preserve_newlines": 5,
    "jslint_happy": false,
    "space_after_anon_function": false,
    "brace_style": "collapse-preserve-inline",
    "keep_array_indentation": true,
    "keep_function_indentation": true,
    "space_before_conditional": true,
    "break_chained_methods": false,
    "eval_code": false,
    "unescape_strings": false,
    "wrap_line_length": 130,
    "wrap_attributes": "auto",
    "wrap_attributes_indent_size": 4,
    "end_with_newline": true,
    "operator_position": "before-newline"
}
```

**Comment by bitwiseman** [comment]
Because of #200, the `// Else do something else` comment is printed before the beautifier knows about the close brace and unindent.  We could solve this one case by look ahead, but it would leave so many others - big bug farm. 

There might be a more general hack/rule, where we special case any group of comments that occur on the lines directly before an element. 

```js
/* block comment */
if (foo) {
    do_something();
    // this comment set has a blank line after it
    // so it stays here
    /* block comments follow the same behavior */

// Else do something else
// this set, has no blank lines after it so it sticks to the element below
} else if (bar) {
    // this set, already behaves correctly
    do_something_else(); 
// this set, has no blank lines after it so it sticks to the element below
/* block comments
 * are not special
 */
// even though there isn't a blank line above.  Seems reasonable.
} else {
    /* still correct */
    do_something_else();  // same line comment stays here

// This is odd, but is what the rule would make happen.  
// Of course, this is also an odd place to put comments
}
next_step();

/* let's talk about unbraced */
if (foo)
    do_something();
// DIFFERENT
// without a brace, this unindents because the if block ended at the semicolon.
// should the example above also follow this rule?  Maybe.

// Else do something else
// this set, has no blank lines after it so it sticks to the element below
else if (bar)
    // this set, already behaves correctly
    do_something_else();
// this set, has no blank lines after it so it sticks to the element below
/* block comments
 * are not special
 */
// even though there isn't a blank line above.  Seems reasonable.
else
    /* still correct */
    do_something_else();  // same line comment stays here

// This no longer looks odd. And is a reasonable place for a comment. 
next_step();
```

**Comment by bitwiseman** [comment]
Gah, the problem is achieving this without better parsing.  So many places that can go wrong. 
Wishful thinking.

**Comment by gagern** [comment]
I disagree with the original request, at least if it is not configurable. I think indented end-of-block comments make a lot of sense in many situations.

```js
if (modulus === 2) {
    // i might be odd here
    i += (i & 1);
    // now i is guaranteed to be even
} else {
    // rounding up using integer arithmetic only
    if (i % modulus)
        i += modulus - (i % modulus);
    // now i is divisible by modulus
}
```

Here I'm using an end-of-block comment to state a post-condition. The comment doesn't directly describe any particular statement or block, but instead describes the state of some object at a specific point during the execution of the code. As such, it should be indented like a statement, not like the closing parenthesis.

If you can find a way to distinguish between these two use cases, that might be fine. I can imagine a global switch somewhere. Making the distinction based on the presence or absence of empty lines somewhere is bound to confuse people, though, so I'd rather avoid that.

**Comment by bitwiseman** [comment]
@gagern - Yeah, this is a particularly difficult problem to land universally. 

I agree with counter-examples. Those should remain where they are. 

I think about it as answering the question "What element is this block of comments attached to?" 
So, we need to define "block of comments" and then what factors we can use to answer what element each block attaches to.   Those factors include, type of previous and following elements, whether a statement/expression ended after the previous or before the following element, and what newlines were present between each block and the previous/following elements. 

No matter what answer we choose it will always before wrong in some cases.  Indentation of other element should not change based on presence or absence of comments except where newlines at the end of comments change semantics.  Beyond that the best we can do is make the behavior predictable and consistent. 

So, what do you think of this:

**Comment by gagern** [comment]
Your example looks good from the point of reading the code, but if code written that way got actually indented to this style, I as a writer of code vould still be very surprised. Personally I'd not expect anything in `{…}` to get dedented, nor would I expect comments trailing an unbraced statement to maintain the indentation. I guess you'd make the distinctions between these cases based on empty lines, but I consider that again to be a source of a lot of confusion.

If you are looking at empty lines, you are essentially looking at the whitespace before beautification. Can't you look at original indentation, too? Note that if the comment is indented like the closing brace, it should stay that way even if both of them now get indented differently. And if the comment is indented like some of the levels of unbraced control constructs, then it can stay with that level. That way, users could control correct indentation simply by indenting correctly, which is a very intuitive approach I think. And js-beautify would know that in some situations several indentations are possible, so it checks all of them in a relative context but picks a preferred one if none of the alternatives matches the input.

**Comment by bitwiseman** [comment]
Using indents is an interesting idea. Thanks!
